ingest/lambdas/indexer.ts

```diff
@@ -0,0 +1,267 @@
+        const bitstreamTextKey = messageData.Records[0].s3.object.key;
+        const [uuid] = bitstreamTextKey.split('.');
+
+        if (!uuid) {
+          throw new Error(`Ingest indexer event missing UUID ${JSON.stringify(event)}`);
+        }
+
+        return {
+          bitstreamTextBucket,
+          bitstreamTextKey,
+          uuid,
+        };
+      }
+    );
+  }
+
+  if (isExplicitInvokeEventSchema(event)) {
+    return [{
+      uuid: event.uuid,
+    }];
+  }
+
+  throw new Error(`Ingest indexer invoked with unknown event: ${JSON.stringify(event)}`);
+};
+
+const index = async (
+  ingestRecord: IngestRecord,
+  dspaceItemBucket: string,
+  openSearchEndpoint: string
+): Promise<void> => {
+  // Get the DSpace item that starts this all off.
+  const dspaceItem = await getDSpaceItemFields(
+    dspaceItemBucket,
+    `${ingestRecord.uuid}.json`
+  );
+
+  // Get the PDF source if it exists.
+  let bitstreamSource: BitstreamTextSource | undefined;
+  if (ingestRecord.bitstreamTextBucket && ingestRecord.bitstreamTextKey) {
+    bitstreamSource = await getBitstreamTextSource(
+      ingestRecord.bitstreamTextBucket,
+      ingestRecord.bitstreamTextKey
+    );
+  }
+
+  // Promote the metadata fields to make search easier. Since the metadata
+  // is dynamic we validate that the dc.title field exists.
+  // TODO: As we add explicit search fields we could turn this into a type.
+  const metadataSearchFields = z.object({ dc_title: z.string() })
+    .passthrough()
+    .parse(getMetadataSearchFields(dspaceItem.metadata));
+
+  // Get the thumbnail retrieve link to make it easier for the UI to render
+  // search results.
+  const thumbnailRetrieveLink = getThumbnailRetrieveLink(
+    dspaceItem.bitstreams
+  );
+
+  // Get terms.
+  const title = metadataSearchFields.dc_title;
+  const contents = bitstreamSource?._bitstreamText || '';
+  const terms = await getTerms(
+    openSearchEndpoint,
+    {
+      title,
+      contents,
+    }
+  );
+
+  const documentItem = documentItemSchema.parse({
+    ...dspaceItem,
+    ...bitstreamSource,
+    ...metadataSearchFields,
+    ...thumbnailRetrieveLink,
+    ...terms,
+  });
+
+  // Index our document item.
+  await osClient.putDocumentItem(openSearchEndpoint, documentItem);
+
+  console.log(`INFO: Indexed document item ${documentItem.uuid}`);
+};
+
+export const handler = async (event: unknown) => {
+  const dspaceItemBucket = getStringFromEnv('DOCUMENT_METADATA_BUCKET');
+  const openSearchEndpoint = getStringFromEnv('OPEN_SEARCH_ENDPOINT');
+
+  const ingestRecords = parseEvent(event);
+
+  // Make sure the documents index exists.
+  const indexExists = await osClient.indexExists(
+    openSearchEndpoint,
+    'documents'
+  );
+
+  // If it doesn't, create it first. If that fails we need to bail because
+  // we can't do anything without the documents index.
+  if (!indexExists) {
+    try {
+      await osClient.createDocumentsIndex(openSearchEndpoint, 'documents');
```

Comment:
There's still a race condition here where the index could be created between the "does it exist" check and the "create it" call. Can we just call "create", catch the error, move along happily if it's an "index already exists" error, and throw any other error?

Reply:
Ah ok that makes sense. Sorry I misunderstood your first comment. I thought you were pointing out the race condition because pMap _could_ have parallelized those requests. I'll fix this now.

Reply:
Fixed.